app/sample_manifest_excel/sample_manifest_excel/upload/processor/tube_rack.rb

```diff
@@ -0,0 +1,201 @@
+          upload.errors.add(:base, error_message)
+          false
+        end
+
+        def validate_coordinates(rack_size, rack_barcode_to_scan_results)
+          rack_layout_map = {
+            48 => {
+              'rows' => 6,
+              'columns' => 8
+            },
+            96 => {
+              'rows' => 8,
+              'columns' => 12
+            }
+          }
+
+          num_rows = rack_layout_map[rack_size]['rows']
+          num_columns = rack_layout_map[rack_size]['columns']
+          valid_row_values = TubeRack.generate_valid_row_values(num_rows)
+          valid_column_values = (1..num_columns).to_a
+
+          rack_barcode_to_scan_results.each_value do |scan_results|
+            scan_results.each_value do |coordinate|
+              row = coordinate.split(//)[0].capitalize
+              column = coordinate.split(//)[1]
+              unless valid_row_values.include?(row) && valid_column_values.include?(column.to_i)
+                error_message = "The coordinate '#{coordinate}' in the scan is not valid for a tube rack of size #{rack_size}."
+                upload.errors.add(:base, error_message)
+                return false
+              end
+            end
+          end
+
+          true
+        end
+
+        def self.generate_valid_row_values(num_rows)
+          output = []
+          count = 1
+          ('A'..'Z').each do |letter|
+            output << letter if count <= num_rows
+            count += 1
+          end
+          output
+        end
+
+        def create_tube_racks_and_link_tubes
+          tube_rack_barcode_to_tube_rack = {}
+          @tube_rack_barcodes_from_manifest.each do |tube_rack_barcode|
+            tube_rack_barcode_to_tube_rack[tube_rack_barcode] = create_tube_rack_if_not_existing(tube_rack_barcode)
+          end
+
+          @upload.rows.each do |row|
+            # create a barcode for the tube
+            tube_barcode = row.value('tube_barcode')
+            tube = row.labware
+            Barcode.create!(asset_id: tube.id, barcode: tube_barcode, format: 7)
+
+            # link the tube to the tube rack
+            tube_rack_barcode = @tube_barcode_to_rack_barcode[tube_barcode]
+            tube_rack = tube_rack_barcode_to_tube_rack[tube_rack_barcode]
+            tube_barcode_to_coordinate = @rack_barcode_to_scan_results[tube_rack_barcode]
+            RackedTube.create!(tube_rack: tube_rack, tube: tube, coordinate: tube_barcode_to_coordinate[tube_barcode])
+          end
+        end
+
+        def create_tube_rack_if_not_existing(tube_rack_barcode)
+          barcode = Barcode.includes(:asset).find_by(asset_id: :tube_rack_barcode)
+
+          if barcode.nil?
+            tube_rack = ::TubeRack.create!(size: @rack_size)
+            Barcode.create!(asset: tube_rack, barcode: tube_rack_barcode, format: 7) # TODO: should we ascertain the format from the barcode, or assume it's fluidx?
```

Comment:
About the TODO comment, my opinion would be to assume the more restrictive (fluidx).

Other people, anymore views about this?